# Exercise 1 -  Sakila Star Schema & ETL  

All the database tables in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120d38_pagila-star/pagila-star.png)

# STEP0: Using ipython-sql

- Load ipython-sql: `%load_ext sql`

- To execute SQL queries you write one of the following atop of your cell: 
    - `%sql`
        - For a one-liner SQL query
        - You can access a python var using `$`    
    - `%%sql`
        - For a multi-line SQL query
        - You can **NOT** access a python var using `$`


- Running a connection string like:
`postgresql://postgres:postgres@db:5432/pagila` connects to the database


# STEP1 : Connect to the local database where Pagila is loaded

##  1.1 Create the pagila db and fill it with data
- Adding `"!"` at the beginning of a jupyter cell runs a command in a shell, i.e. we are not running python code but we are running the `createdb` and `psql` postgresql commmand-line utilities

In [19]:
%load_ext autoreload
%autoreload 2

In [16]:
import os
import psycopg2

In [14]:
current_dir = os.getcwd()
data_path = os.path.join(os.path.join(current_dir, '..'), 'Data')
data_path

'/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_data_warehouses/exercises/../Data'

In [15]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f {data_path}/pagila-schema.sql
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f {data_path}/pagila-data.sql

createdb: error: database creation failed: ERROR:  database "pagila" already exists
psql:/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_data_warehouses/exercises/../Data/pagila-schema.sql:28: ERROR:  must be owner of extension plpgsql
psql:/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_data_warehouses/exercises/../Data/pagila-schema.sql:46: ERROR:  must be member of role "postgres"
psql:/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_data_warehouses/exercises/../Data/pagila-schema.sql:56: ERROR:  must be member of role "postgres"
psql:/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_data_warehouses/exercises/../Data/pagila-schema.sql:73: ERROR:  must be member of role "postgres"
psql:/Users/adamarnon/Desktop/learning/udacity/udacity-data-eng-aws/2_Cloud_Data_Warehouses/l1_intro_to_dat

## 1.2 Connect to the newly created db

In [20]:
%load_ext sql

In [21]:
DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)


postgresql://student:student@127.0.0.1:5432/pagila


In [22]:
%sql $conn_string

# STEP2 : Explore the  3NF Schema

<img src="../pagila-3nf.png" width="50%"/>

## 2.1 How much? What data sizes are we looking at?

In [23]:
nStores = %sql select count(*) from store;
nFilms = %sql select count(*) from film;
nCustomers = %sql select count(*) from customer;
nRentals = %sql select count(*) from rental;
nPayment = %sql select count(*) from payment;
nStaff = %sql select count(*) from staff;
nCity = %sql select count(*) from city;
nCountry = %sql select count(*) from country;

print("nFilms\t\t=", nFilms[0][0])
print("nCustomers\t=", nCustomers[0][0])
print("nRentals\t=", nRentals[0][0])
print("nPayment\t=", nPayment[0][0])
print("nStaff\t\t=", nStaff[0][0])
print("nStores\t\t=", nStores[0][0])
print("nCities\t\t=", nCity[0][0])
print("nCountry\t\t=", nCountry[0][0])

 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
nFilms		= 1000
nCustomers	= 599
nRentals	= 16044
nPayment	= 16049
nStaff		= 2
nStores		= 2
nCities		= 600
nCountry		= 109


## 2.2 When? What time period are we talking about?

In [24]:
%%sql 
select min(payment_date) as start, max(payment_date) as end from payment;

 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.


[(datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), datetime.datetime(2017, 5, 14, 13, 44, 29, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=61200))))]

## 2.3 Where? Where do events in this database occur?
TODO: Write a query that displays the number of addresses by district in the address table. Limit the table to the top 10 districts. Your results should match the table below.

In [43]:
%%sql
select district, count(address) as n
from address
group by district
order by n desc
limit 10;


 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('Buenos Aires', 10),
 ('Shandong', 9),
 ('California', 9),
 ('West Bengali', 9),
 ('Uttar Pradesh', 8),
 ('So Paulo', 8),
 ('England', 7),
 ('Maharashtra', 7),
 ('Southern Tagalog', 6),
 ('Gois', 5)]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>district</th>
        <th>n</th>
    </tr>
    <tr>
        <td>Shandong</td>
        <td>3237</td>
    </tr>
    <tr>
        <td>England</td>
        <td>2974</td>
    </tr>
    <tr>
        <td>So Paulo</td>
        <td>2952</td>
    </tr>
    <tr>
        <td>West Bengali</td>
        <td>2623</td>
    </tr>
    <tr>
        <td>Buenos Aires</td>
        <td>2572</td>
    </tr>
    <tr>
        <td>Uttar Pradesh</td>
        <td>2462</td>
    </tr>
    <tr>
        <td>California</td>
        <td>2444</td>
    </tr>
    <tr>
        <td>Southern Tagalog</td>
        <td>1931</td>
    </tr>
    <tr>
        <td>Tamil Nadu</td>
        <td>1807</td>
    </tr>
    <tr>
        <td>Hubei</td>
        <td>1790</td>
    </tr>
</tbody></table></div>

# STEP3: Perform some simple data analysis

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed steps 1 and 2, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

In [ ]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

### 3NF - Entity Relationship Diagram

<img src="../pagila-3nf.png" width="50%"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [44]:
%%sql
select film_id, title, release_year, rental_rate, rating  from film limit 5;


 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2006, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2006, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2006, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2006, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2006, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [45]:
%%sql
select * from payment limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(16050, 269, 2, 7, Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (16052, 269, 2, 678, Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (16053, 269, 2, 703, Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (16054, 269, 1, 750, Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))))]

### 3.1.3 Inventory

In [46]:
%%sql
select * from inventory limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (2, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (3, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (4, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600)))),
 (5, 1, 2, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))))]

### 3.1.4 Get the movie of every payment

In [47]:
%%sql
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[('SWARM GOLD', Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), 269),
 ('PACKER MADIGAN', Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), 269),
 ('SOMETHING DUCK', Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), 269),
 ('DRACULA CRYSTAL', Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), 269),
 ('CLOSER BANG', Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), 269)]

### 3.1.5 sum movie rental revenue
TODO: Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [52]:
%%sql
select title, sum(amount) as revenue
from (
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
 ) as movie_payment
group by title
order by revenue desc
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('231.73')),
 ('WIFE TURN', Decimal('223.69')),
 ('ZORRO ARK', Decimal('214.69')),
 ('GOODFELLAS SALUTE', Decimal('209.69')),
 ('SATURDAY LAMBS', Decimal('204.72')),
 ('TITANS JERK', Decimal('201.71')),
 ('TORQUE BOUND', Decimal('198.72')),
 ('HARRY IDAHO', Decimal('195.70')),
 ('INNOCENT USUAL', Decimal('191.74')),
 ('HUSTLER PARTY', Decimal('190.78'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [53]:
%%sql
SELECT p.customer_id, p.rental_id, p.amount, ci.city                            
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[(130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (408, 3, Decimal('3.99'), 'Jaffna'),
 (333, 4, Decimal('4.99'), 'Baku'),
 (222, 5, Decimal('6.99'), 'Jaroslavl'),
 (549, 6, Decimal('0.99'), 'Santiago de Compostela'),
 (269, 7, Decimal('1.99'), 'Salinas'),
 (239, 8, Decimal('4.99'), 'Ciomas'),
 (126, 9, Decimal('4.99'), 'Po'),
 (399, 10, Decimal('5.99'), 'Okara')]

### 3.2.2 Top grossing cities
TODO: Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [54]:
%%sql
SELECT  city, sum(amount) as revenue
from (
    SELECT p.customer_id, p.rental_id, p.amount, ci.city                            
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
) as customer_payment
group by city
order by revenue desc
limit 10;


 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('Cape Coral', Decimal('221.55')),
 ('Saint-Denis', Decimal('216.54')),
 ('Aurora', Decimal('198.50')),
 ('Molodetno', Decimal('195.58')),
 ('Santa Brbara dOeste', Decimal('194.61')),
 ('Apeldoorn', Decimal('194.61')),
 ('Qomsheh', Decimal('186.62')),
 ('London', Decimal('180.52')),
 ('Ourense (Orense)', Decimal('177.60')),
 ('Bijapur', Decimal('175.61'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [55]:
%%sql
SELECT sum(p.amount) as revenue, EXTRACT(month FROM p.payment_date) as month
from payment p
group by month
order by revenue desc
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(Decimal('28559.46'), Decimal('4')),
 (Decimal('23886.56'), Decimal('3')),
 (Decimal('9631.88'), Decimal('2')),
 (Decimal('4824.43'), Decimal('1')),
 (Decimal('514.18'), Decimal('5'))]

### 3.3.2 Each movie by customer city and by month (data cube)

In [56]:
%%sql
SELECT f.title, p.amount, p.customer_id, ci.city, p.payment_date,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('GRADUATE LORD', Decimal('3.99'), 408, 'Jaffna', datetime.datetime(2017, 1, 24, 21, 32, 5, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('LOVE SUICIDES', Decimal('4.99'), 333, 'Baku', datetime.datetime(2017, 1, 24, 21, 33, 7, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('IDOLS SNATCHERS', Decimal('6.99'), 222, 'Jaroslavl', datetime.datetime(2017, 1, 24, 21, 33, 47, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('MYSTIC TRUMAN', Decimal('0.99'), 549, 'Santiago de Compostela', datetime.datetime(2017, 1, 24, 21, 36, 33, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('SWARM GOLD', Decimal('1.99'), 269, 'Salinas', datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('LAWLESS VISION', Decimal('4.99'), 239, 'Ciomas', datetime.datetime(2017, 1, 24, 22, 0, 12, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('MATRIX SNOWMAN', Decimal('4.99'), 126, 'Po', datetime.datetime(2017, 1, 24, 22, 29, 6, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1')),
 ('HANGING DEEP', Decimal('5.99'), 399, 'Okara', datetime.datetime(2017, 1, 24, 22, 30, 47, 996577, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=57600))), Decimal('1'))]

### 3.3.3 Sum of revenue of each movie by customer city and by month

TODO: Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [63]:
%%sql
SELECT f.title, ci.city,EXTRACT(month FROM p.payment_date) as month, sum(p.amount) as revenue
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
group by (f.title, ci.city, month)
order by month, revenue desc
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('SHOW LORD', 'Mannheim', Decimal('1'), Decimal('11.99')),
 ('KISSING DOLLS', 'Toulon', Decimal('1'), Decimal('10.99')),
 ('TELEGRAPH VOYAGE', 'Naala-Porto', Decimal('1'), Decimal('10.99')),
 ('AMERICAN CIRCUS', 'Callao', Decimal('1'), Decimal('10.99')),
 ('CASUALTIES ENCINO', 'Warren', Decimal('1'), Decimal('10.99')),
 ('AUTUMN CROW', 'Ashgabat', Decimal('1'), Decimal('9.99')),
 ('MIDSUMMER GROUNDHOG', 'Vaduz', Decimal('1'), Decimal('9.99')),
 ('DAY UNFAITHFUL', 'Baybay', Decimal('1'), Decimal('9.99')),
 ('CALIFORNIA BIRDS', 'Plock', Decimal('1'), Decimal('9.99')),
 ('DARKO DORADO', 'Bhilwara', Decimal('1'), Decimal('9.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>